File: threads/src/main/java/im/threads/internal/model/Attachment.java
package im.threads.internal.model;

import android.text.TextUtils;

public final class Attachment {
    private String result;
    private boolean isSelfie;
    private Optional optional;
    private AttachmentStateEnum state = AttachmentStateEnum.ANY;
    private ErrorStateEnum errorCode = ErrorStateEnum.ANY;
    private String errorMessage = "";
    private String name;
    private String type;

    public String getResult() {
        return result;
    }

    public void setResult(String result) {
        this.result = result;
    }

    public Optional getOptional() {
        return optional;
    }

    public void setOptional(Optional optional) {
        this.optional = optional;
    }

    public boolean isSelfie() {
        return isSelfie;
    }

    public void setSelfie(boolean selfie) {
        isSelfie = selfie;
    }

    public AttachmentStateEnum getState() {
        return state;
    }

    public void setState(AttachmentStateEnum state) {
        this.state = state;
    }

    public ErrorStateEnum getErrorCode() {
        return errorCode;
    }

    public void setErrorCode(ErrorStateEnum errorCode) {
        this.errorCode = errorCode;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public void setErrorMessage(String errorMessage) {
        this.errorMessage = errorMessage;
    }

    public String getName() {
        if(!TextUtils.isEmpty(name))
            return name;
        if(optional != null)
            return optional.getName();
        return null;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getType() {
        if(!TextUtils.isEmpty(type))
            return type;
        if(optional != null)
            return optional.getType();
        return null;
    }

    public void setType(String type) {
        this.type = type;
    }
}

File: threads/src/main/java/im/threads/internal/model/Attachment.kt
package im.threads.internal.model

import im.threads.internal.model.ErrorStateEnum.Companion.errorStateEnumFromString

class Attachment {
    var result: String? = null
    var optional: Optional? = null
    var state = AttachmentStateEnum.ANY
    var errorCode: String? = null
    var errorMessage = ""

    var name: String? = null
        get() {
            return if (!field.isNullOrEmpty()) field
            else optional?.name
        }

    var type: String? = null
        get() {
            return if (!field.isNullOrEmpty()) field
            else optional?.type
        }

    fun getErrorCodeState(): ErrorStateEnum {
        errorCode?.let {
            return errorStateEnumFromString(it)
        }
        return ErrorStateEnum.ANY
    }
}
